File: Tetrix/Project/src/hidnif.erl
-module(hidnif).
-export([start_usb/0,stop_usb/1,set_speed/2,set_angle/2,set_display/2,set_brakeLight/2,
set_leftLight/2,set_rightLight/2,start/0,loop/2,get_accelSpeed/1,get_modeSwitch/1,get_remoteStatus/1,
get_Voltage/1,get_Current/1,get_Heading/1,get_IR0/1,get_IR1/1,get_ultraSonic/1]).


start_usb() ->
    exit("NIF library not loaded").

stop_usb(_Address) ->
    exit("NIF library not loaded").

set_speed(_Speed, _Address) ->
    exit("NIF library not loaded").

set_angle(_Angle, _Address) ->
    exit("NIF library not loaded").

set_display(_Text, _Address) ->
    exit("NIF library not loaded").

set_brakeLight(_bValue, _Address) ->
    exit("NIF library not loaded").

set_leftLight(_lValue, _Address) ->
    exit("NIF library not loaded").

set_rightLight(_rValue, _Address) ->
    exit("NIF library not loaded").

get_accelSpeed(_Address) ->
    exit("NIF library not loaded"). 

get_modeSwitch(_Address) ->
    exit("NIF library not loaded"). 

get_remoteStatus(_Address) ->
    exit("NIF library not loaded"). 

get_Voltage(_Address) ->
    exit("NIF library not loaded"). 

get_Current(_Address) ->
    exit("NIF library not loaded"). 

get_Heading(_Address) ->
    exit("NIF library not loaded").

get_IR0(_Address) ->
    exit("NIF library not loaded").

get_IR1(_Address) ->
    exit("NIF library not loaded").

get_ultraSonic(_Address) ->
    exit("NIF library not loaded").

start() ->
    erlang:load_nif("./hidnif", 0),
    Adress=start_usb(),
    Pid=self(),
    register(?MODULE, spawn(hidnif,loop,[Adress,Pid])).

loop(Adress,Pid)->
    receive
	{setSpeed,Value}->
	    set_speed(Value, Adress), 
            loop(Adress,Pid);
	{setAngle,Value}->
	    set_angle(Value, Adress), 
            loop(Adress,Pid);
	{setDisplay,Value}->
	    set_display(Value, Adress), 
            loop(Adress,Pid);
	{setBrakeLight,Value}->
	    set_brakeLight(Value, Adress), 
            loop(Adress,Pid);
	{setLeftLight,Value}->
	    set_leftLight(Value, Adress), 
            loop(Adress,Pid);
	{setRightLight,Value}->
	    set_rightLight(Value, Adress), 
            loop(Adress,Pid);
	{getAccelSpeed}->
	    Tmp=get_accelSpeed(Adress),
	    Pid ! {curSpeed,Tmp}, 
            loop(accelSpeed,Pid);
	{getModeSwitch}->
	    Tmp=get_modeSwitch(Adress),
	    Pid ! {modeSwitch,Tmp}, 
            loop(Adress,Pid);
	{getRemoteStatus}->
	    Tmp=get_remoteStatus(Adress),
	    Pid ! {remoteStatus,Tmp}, 
            loop(Adress,Pid);
	{getVoltage}->
	    Tmp=get_Voltage(Adress),
            Pid ! {voltage,Tmp}, 
            loop(Adress,Pid);
	{getCurrent}->
	    Tmp=get_Current(Adress),
	    Pid ! {current,Tmp}, 
            loop(Adress,Pid);
	{getHeading}->
	    Tmp=get_Heading(Adress),
            Pid ! {heading,Tmp}, 
            loop(Adress,Pid);
	{getIR0}->
	    Tmp=get_IR0(Adress),
            Pid ! {iR0,Tmp}, 
            loop(Adress,Pid);
	{getIR1}->
	    Tmp=get_IR1(Adress),
            Pid ! {iR1,Tmp}, 
            loop(Adress,Pid);
	{getUltraSonic}->
	    Tmp=get_ultraSonic(Adress),
            Pid ! {ultraSonic,Tmp}, 
            loop(Adress,Pid);
        {stop}->
	    stop_usb(Adress)
    end.
	

File: 2014-CaroloCup/Tetrix/Project/c_source/hidnif.cpp
#include "erl_nif.h"
#include <sys/time.h>
#include "iostream"
#include <stdio.h>
#include <wchar.h>
#include <string.h>
#include "../c_include/carControl.h"



using namespace std;

extern void setAngle(int val,long mem);
extern void setSpeed(int val,long mem);



static ERL_NIF_TERM start_usb(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  address = startUsb();
  return enif_make_long(env, address);
}
static ERL_NIF_TERM stop_usb(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  stopUsb(address);
  return enif_make_atom(env,"ok");
}

static ERL_NIF_TERM set_angle(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  int angle;
  long address;
  enif_get_int(env, argv[0], &angle);
  enif_get_long(env, argv[1], &address);
  setAngle(angle, address);
  return enif_make_long(env,address);
}

static ERL_NIF_TERM set_speed(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  int speed;
  long address;
  enif_get_int(env, argv[0], &speed);
  enif_get_long(env, argv[1], &address);
  setSpeed(speed, address);
  return enif_make_long(env,address);
}



static ERL_NIF_TERM set_display(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  char display[32];
  long address;
  enif_get_string(env, argv[0], display, 32, ERL_NIF_LATIN1);
  enif_get_long(env, argv[1], &address);
  setDisplay(display, address);
  return enif_make_long(env,address);
}

static ERL_NIF_TERM set_BrakeLight(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  int brakeL;
  long address;
  enif_get_int(env, argv[0], &brakeL);
  enif_get_long(env, argv[1], &address);
  setBrakeLight(brakeL, address);
  return enif_make_long(env,address);
}

static ERL_NIF_TERM set_RightSignalLight(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  int rightL;
  long address;
  enif_get_int(env, argv[0], &rightL);
  enif_get_long(env, argv[1], &address);
  setRightSignalLight(rightL, address);
  return enif_make_long(env,address);
}

static ERL_NIF_TERM set_LeftSignalLight(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  int leftL;
  long address;
  enif_get_int(env, argv[0], &leftL);
  enif_get_long(env, argv[1], &address);
  setLeftSignalLight(leftL, address);
  return enif_make_long(env,address);
}

static ERL_NIF_TERM get_accelSpeed(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_double(env,getAxelSpeed(address));
}

static ERL_NIF_TERM get_modeSwitch(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_int(env,getModeSwitch(address));
}

static ERL_NIF_TERM get_remoteStatus(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_int(env,getRemoteStatus(address));
}

static ERL_NIF_TERM get_Voltage(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_double(env,getVoltage(address));
}

static ERL_NIF_TERM get_Current(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_double(env,getCurrent(address));
}

static ERL_NIF_TERM get_Heading(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_int(env,getHeading(address));
}

static ERL_NIF_TERM get_IR0(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_double(env,getIrSensor0(address));
}

static ERL_NIF_TERM get_IR1(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_double(env,getIrSensor1(address));
}

static ERL_NIF_TERM get_ultraSonic(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  long address;
  enif_get_long(env, argv[0], &address);
  return enif_make_int(env,getUltraSonic(address));
}

/*
For testing purposes, crash the system with a segmentation fault
*/
static ERL_NIF_TERM cause_segmentation_fault(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
  char *s = "hello world";
  *s = 'H';  
  return enif_make_double(env,5);
}

static ErlNifFunc nif_funcs[] = 
  {
    
    {"start_usb", 0, start_usb},
    {"stop_usb", 1, stop_usb},
    {"set_speed", 2, set_speed},
    {"set_angle", 2, set_angle}, 
    {"set_display", 2, set_display},
    {"set_brakeLight", 2, set_BrakeLight},
    {"set_rightLight", 2, set_RightSignalLight},
    {"set_leftLight", 2, set_LeftSignalLight},
    {"get_accelSpeed", 1, get_accelSpeed},
    {"get_modeSwitch", 1, get_modeSwitch},
    {"get_remoteStatus", 1, get_remoteStatus},
    {"get_Voltage", 1, get_Voltage},
    {"get_Current", 1, get_Current},
    {"get_Heading", 1, get_Heading},
    {"get_IR0", 1, get_IR0},
    {"get_IR1", 1, get_IR1},
    {"get_ultraSonic", 1, get_ultraSonic},
    {"cause_segmentation_fault",0, cause_segmentation_fault}
  };

ERL_NIF_INIT(hidnif,nif_funcs,NULL,NULL,NULL,NULL)
